private/pkg/storage/bucket.go

```diff
@@ -52,6 +52,38 @@ type ReadBucket interface {
 	Walk(ctx context.Context, prefix string, f func(ObjectInfo) error) error
 }
 
+// PutOptions are the possible options that can be passed to a Put operation.
+type PutOptions struct {
+	IfNotExists      bool
+	CustomChunkSize  bool
+	ChunkSizeInBytes int
+}
+
+// PutOption are options passed when putting an object in a bucket.
+type PutOption func(*PutOptions)
+
+// PutIfNotExists only puts the object in the bucket if there is no other object
+// already in the same bucket path. Some implementations of
+// `storage.WriteBucket.Put` that are `ReadWriteBucket` support this option.
+// Check the implementation of the bucket you're using if you're passing this
+// option.
+func PutIfNotExists() PutOption {
+	return func(opts *PutOptions) {
+		opts.IfNotExists = true
+	}
+}
+
+// PutWithChunkBytesSize sets a custom chunk size in bytes for writing the
+// object. Some implementations of `storage.WriteBucket.Put` allow multi-part
+// upload, and allow customizing the chunk size of each part upload. Check the
```

Comment:
What does "check the implementation" mean? Callers of `storage` types should not have to concern themselves with what the implementation is at all, and all options should be defined in such a way that all implementations can reliably implement them. This is a core tenet of `storage` - callers only need to know about the API, not the implementation (really a core of all APIs). This may mean that this option is defined as "this is a suggested chunk size, implementations may choose to ignore this option".

Reply:
You could do something similar to `SetExternalPathSupported` as well, but in this case, that likely is not preferred

Reply:
Yeah, @pkwarren was suggesting to also have an `UnsupportedOption` error top level here, so implementations can return it instead of ignoring a given option.